lib/rules/no-inner-declarations.js

```diff
@@ -66,6 +76,12 @@ module.exports = {
 
     create(context) {
 
+        const sourceCode = context.sourceCode;
+        const ecmaVersion = context.languageOptions.ecmaVersion ?? 6;
+        const legacy = context.options[1]?.legacy ?? false;
+
+        const isStrict = sourceCode.getScope(sourceCode.ast).isStrict;
```

Comment:
We should check the scope where the function is declared.

For example, this whole script is non-strict code so the legacy behavior that function declarations are not block-scoped applies:

```js
// script

function foo() {
    {
        function bar() {
            console.log("bar called");
        }
    }
    bar();
}

foo(); // logs "bar called"

```

But, for function declarations to be block-scoped, the whole script doesn't need to be strict mode code, just the code where the function is declared needs to be strict mode code.

```js
// script

function foo() {
    "use strict"; // <-
    {
        function bar() {
            console.log("bar called");
        }
    }
    bar();
}

foo(); // ReferenceError: bar is not defined

```